File: core/server/routers/nodes.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from core.server.database import get_db
from core.server.models import Node
from core.shared.schemas import NodeInfo
from datetime import datetime, timezone

router = APIRouter()

@router.post("/register", status_code=status.HTTP_200_OK)
def register_node(node_info: NodeInfo, db: Session = Depends(get_db)):
    db_node = db.query(Node).filter(Node.node_id == node_info.node_id).first()
    if db_node:
        db_node.os_name = node_info.os_name
        db_node.ip_address = node_info.ip_address
        db_node.openclaw_version = node_info.openclaw_version
        db_node.last_seen_at = datetime.now(timezone.utc)
    else:
        db_node = Node(
            node_id=node_info.node_id,
            os_name=node_info.os_name,
            ip_address=node_info.ip_address,
            openclaw_version=node_info.openclaw_version,
            last_seen_at=datetime.now(timezone.utc)
        )
        db.add(db_node)
    db.commit()
    return {"message": "Node registered successfully", "node_id": node_info.node_id}

@router.get("/list")
def list_nodes(db: Session = Depends(get_db)):
    nodes = db.query(Node).order_by(Node.last_seen_at.desc()).all()
    return [
        {
            "node_id": n.node_id,
            "os_name": n.os_name,
            "ip_address": n.ip_address,
            "openclaw_version": n.openclaw_version,
            "last_seen_at": n.last_seen_at.isoformat() if n.last_seen_at else None,
        }
        for n in nodes
    ]

File: services/server/api/nodes.go
package api

import (
	"time"

	"github.com/gin-gonic/gin"

	"github.com/danl5/clawo11y/services/server/database"
	"github.com/danl5/clawo11y/services/server/models"
	"github.com/danl5/clawo11y/services/server/schemas"
)

func RegisterNode(c *gin.Context) {
	var payload schemas.NodeInfo
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(400, gin.H{"detail": err.Error()})
		return
	}

	openclawVersion := ""
	if payload.OpenClawVersion != nil {
		openclawVersion = *payload.OpenClawVersion
	}
	hostname := ""
	if payload.Hostname != nil {
		hostname = *payload.Hostname
	}

	dbNode := models.Node{
		NodeID:          payload.NodeID,
		OSName:          payload.OSName,
		IPAddress:       payload.IPAddress,
		OpenClawVersion: openclawVersion,
		Hostname:        hostname,
		LastSeenAt:      time.Now(),
	}

	// Upsert node (SQLite compatible way)
	var existingNode models.Node
	result := database.DB.Where("node_id = ?", payload.NodeID).First(&existingNode)

	if result.Error == nil {
		// Found, update it
		existingNode.OSName = payload.OSName
		existingNode.IPAddress = payload.IPAddress
		existingNode.OpenClawVersion = openclawVersion
		existingNode.Hostname = hostname
		existingNode.LastSeenAt = time.Now()
		if err := database.DB.Save(&existingNode).Error; err != nil {
			c.JSON(500, gin.H{"detail": err.Error()})
			return
		}
	} else {
		// Not found, create it
		if err := database.DB.Create(&dbNode).Error; err != nil {
			c.JSON(500, gin.H{"detail": err.Error()})
			return
		}
	}

	c.JSON(200, gin.H{"message": "Node registered successfully", "node_id": payload.NodeID})
}

func GetNodes(c *gin.Context) {
	var nodes []models.Node
	if err := database.DB.Find(&nodes).Error; err != nil {
		c.JSON(500, gin.H{"detail": err.Error()})
		return
	}

	// Format response to match python
	resp := make([]gin.H, 0, len(nodes))
	for _, n := range nodes {
		resp = append(resp, gin.H{
			"node_id":          n.NodeID,
			"os_name":          n.OSName,
			"ip_address":       n.IPAddress,
			"openclaw_version": n.OpenClawVersion,
			"hostname":         n.Hostname,
			"last_seen_at":     n.LastSeenAt.Format(time.RFC3339Nano),
		})
	}
	c.JSON(200, resp)
}
